src/soc/ibm/power9/istep_10_1.c

```diff
@@ -0,0 +1,840 @@
+					  uint64_t data)
+{
+	for (uint8_t i = 0; i < P9_BUILD_SMP_NUM_SHADOWS; i++)
+		put_scom(chip, shadow_regs[i], data);
+}
+
+static void p9_build_smp_copy_hp_ab_next_curr(uint8_t chip)
+{
+	/* Read NEXT */
+	uint64_t hp_mode_data = p9_build_smp_get_hp_ab_shadow(chip, PB_HP_MODE_NEXT_SHADOWS);
+	uint64_t hpx_mode_data = p9_build_smp_get_hp_ab_shadow(chip, PB_HPX_MODE_NEXT_SHADOWS);
+	uint64_t hpa_mode_data = p9_build_smp_get_hp_ab_shadow(chip, PB_HPA_MODE_NEXT_SHADOWS);
+
+	/* Write CURR */
+	p9_build_smp_set_hp_ab_shadow(chip, PB_HP_MODE_CURR_SHADOWS, hp_mode_data);
+	p9_build_smp_set_hp_ab_shadow(chip, PB_HPX_MODE_CURR_SHADOWS, hpx_mode_data);
+	p9_build_smp_set_hp_ab_shadow(chip, PB_HPA_MODE_CURR_SHADOWS, hpa_mode_data);
+}
+
+static void p9_build_smp_copy_hp_ab_curr_next(uint8_t chip)
+{
+	/* Read CURR */
+	uint64_t hp_mode_data  = p9_build_smp_get_hp_ab_shadow(chip, PB_HP_MODE_CURR_SHADOWS);
+	uint64_t hpx_mode_data = p9_build_smp_get_hp_ab_shadow(chip, PB_HPX_MODE_CURR_SHADOWS);
+	uint64_t hpa_mode_data = p9_build_smp_get_hp_ab_shadow(chip, PB_HPA_MODE_CURR_SHADOWS);
+
+	/* Write NEXT */
+	p9_build_smp_set_hp_ab_shadow(chip, PB_HP_MODE_NEXT_SHADOWS, hp_mode_data);
+	p9_build_smp_set_hp_ab_shadow(chip, PB_HPX_MODE_NEXT_SHADOWS, hpx_mode_data);
+	p9_build_smp_set_hp_ab_shadow(chip, PB_HPA_MODE_NEXT_SHADOWS, hpa_mode_data);
+}
+
+static void p9_build_smp_set_fbc_ab(void)
+{
+	/*
+	 * quiesce 'slave' fabrics in preparation for joining
+	 *   PHASE1 -> quiesce all chips except the chip which is the new fabric master
+	 *   PHASE2 -> quiesce all drawers except the drawer containing the new fabric master
+	 */
+	p9_build_smp_sequence_adu(QUIESCE);
+
+	/* Program NEXT register set for all chips via initfile */
+	p9_fbc_ab_hp_scom(/*chip=*/0);
+	p9_fbc_ab_hp_scom(/*chip=*/1);
+
+	/* Program CURR register set only for chips which were just quiesced */
+	p9_build_smp_copy_hp_ab_next_curr(/*chip=*/1);
+
+	/*
+	 * Issue switch AB reconfiguration from chip designated as new master
+	 * (which is guaranteed to be a master now)
+	 */
+	p9_build_smp_sequence_adu(SWITCH_AB);
+
+	/* Reset NEXT register set (copy CURR->NEXT) for all chips */
+	p9_build_smp_copy_hp_ab_curr_next(/*chip=*/0);
+	p9_build_smp_copy_hp_ab_curr_next(/*chip=*/1);
+}
+
+static void p9_build_smp(void)
+{
+	/* Apply three CD hotplug sequences to each chip to initialize SCOM
+	 * chains */
+	for (int seq = 1; seq <= 3; seq++) {
+		for (int chip = 0; chip < 2; chip++) {
+			if (seq == 1)
+				p9_fbc_cd_hp1_scom(chip);
+			else
+				p9_fbc_cd_hp23_scom(chip, seq);
+		}
+
+		/* Issue switch CD on all chips to force updates to occur */
+		p9_build_smp_sequence_adu(SWITCH_CD);
+	}
+
+	p9_build_smp_set_fbc_ab();
+}
+
+void istep_10_1(uint8_t chips)
+{
+	printk(BIOS_EMERG, "starting istep 10.1\n");
+	report_istep(10,1);
+
+	if (chips != 0x01) {
+		p9_build_smp();
+
+		/* Sanity check that XSCOM works for the second CPU */
+		if (read_rscom(1, 0xf000f) != 0x223d104900808040)
```

Comment:
This will have to be changed to work with anything else than our CPU model. I don't know if it is worth to check for expected values (e.g. VENDOR_ID and CHIPPOS_ID), or if testing for not 0 and not 0xFF... is enough.

Same goes for similar test in romstage.c - I believe I fixed that one already but can't recall on which branch and how.